src/ConstructionBase.jl

```diff
@@ -24,129 +24,145 @@ for (name, path) in [
     end
 end
 
-@generated function constructorof(::Type{T}) where T
-    getfield(parentmodule(T), nameof(T))
+@static if isdefined(Base, Symbol("@assume_effects"))
+    using Base: @assume_effects
+else
+    macro assume_effects(_, ex)
+        Base.@pure ex
+    end
 end
+struct PropertyNames{syms} end
```

Comment:
I like the idea of statically wrapping property names in a struct. Is this for ergonomics, e.g. `getproperties` on a subset of properties or does this improve inference?

Reply:
A bit of both. `propertynames` isn't necessarily inferrible at compile time and `fieldnames` is a bit fickle still. It's also safer to use strictly known parameters in `@generated` functions.

Reply:
ok thanks!